src/__tests__/databaseTests.js

```diff
@@ -1,858 +1,877 @@
 
-export const matchTests = [
-  {
-    name: 'matches everything',
-    query: [],
-    matching: [{ id: 'm1' }, { id: 'm2', num1: 10 }],
-    nonMatching: [],
-  },
-  {
-    name: 'matches strings',
-    query: [Q.where('text1', 'value1')],
-    matching: [{ id: 'm1', text1: 'value1' }],
-    nonMatching: [{ id: 'n1' }, { id: 'n2', text1: null }, { id: 'n3', text1: 'other_value' }],
-  },
-  {
-    name: 'matches `true`',
-    query: [Q.where('bool1', true)],
-    matching: [
-      { id: 'm1', bool1: true },
-      { id: 'm2', bool1: 1 },
-    ],
-    nonMatching: [
-      { id: 'n1' },
-      { id: 'n2', bool1: null },
-      { id: 'n3', bool1: false },
-      { id: 'n4', bool1: 0 },
-    ],
-  },
-  {
-    name: 'matches `false`',
-    query: [Q.where('bool2', false)],
-    matching: [
-      { id: 'm1', bool2: false },
-      { id: 'm2', bool2: 0 },
-    ],
-    nonMatching: [
-      { id: 'n1' },
-      { id: 'n2', bool2: null },
-      { id: 'n3', bool2: true },
-      { id: 'n4', bool2: 1 },
-    ],
-  },
-  {
-    name: 'matches `null`',
-    query: [Q.where('num1', null)],
-    matching: [{ id: 'm1' }, { id: 'm2', num1: null }, { id: 'm3', num1: undefined }],
-    nonMatching: [
-      { id: 'n1', num1: 0 },
-      { id: 'n2', num1: false },
-      { id: 'n3', num1: '' },
-    ],
-  },
-  {
-    name: 'matches integers (0)',
-    query: [Q.where('num1', 0)],
-    matching: [
-      { id: 'm1', num1: 0 },
-      { id: 'm2', num1: false },
-    ],
-    nonMatching: [
-      { id: 'n1' },
-      { id: 'n2', num1: null },
-      { id: 'n3', num1: 1 },
-      { id: 'n4', num1: true },
-    ],
-  },
-  {
-    name: 'matches integers (1)',
-    query: [Q.where('num1', 1)],
-    matching: [
-      { id: 'm1', num1: 1 },
-      { id: 'm2', num1: true },
-    ],
-    nonMatching: [
-      { id: 'n1' },
-      { id: 'n2', num1: null },
-      { id: 'n3', num1: 0 },
-      { id: 'n4', num1: false },
-    ],
-  },
-  {
-    name: 'matches floats',
-    query: [Q.where('float1', 3.14)],
-    matching: [{ id: 'm1', float1: 3.14 }],
-    nonMatching: [
-      { id: 'n1', float1: null },
-      { id: 'n2', float1: 1.0 },
-    ],
-  },
-  {
-    name: 'matches big numbers (e.g. JS timestamps)',
-    query: [Q.where('float1', 1590485104033)],
-    matching: [{ id: 'm1', float1: 1590485104033 }],
-    nonMatching: [
-      { id: 'n1', float1: null },
-      { id: 'n2', float1: 159048510 },
-    ],
-  },
-  {
-    name: 'matches multiple conditions',
-    query: [
-      Q.where('text1', `value1`),
-      Q.where('num1', 2),
+const Q = (QueryBuilders: any)
+
+const matchTests: any[] = []
+
+function matchTest(
+  options: $Exact<{
+    name: string,
+    query: QueryBuilders.Clause[],
+    matching: Array<{ id: string, ... }>,
+    nonMatching: Array<{ id: string, ... }>,
+    skipQuery?: boolean,
+    skipCount?: boolean,
+    skipLoki?: boolean,
+    skipSqlite?: boolean,
+    skipMatcher?: boolean,
+  }>,
+): void {
+  matchTests.push(options)
```

Comment:
Why is `matchTest` and `joinTest` needed? Wouldn't providing an exact type for `matchTests` and `joinTests` instead of `any` do the job?

Reply:
it's not _really_ needed, it's just that joinTests are skipped for `encodeMatcher` tests, since they're all incompatible. I can refactor it but doesn't really matter IMO ¯\_(ツ)_/¯  I don't want to spend time typing out tests, I only enabled flow in this one particular file to catch most obvious errors - missing fields for the tests and such